sympy/solvers/solveset.py

```diff
@@ -789,16 +785,29 @@ def _solveset(f, symbol, domain, _check=False):
         a = f.args[0]
         result = solveset_real(a > 0, symbol)
     elif f.is_Piecewise:
-        dom = domain
         result = EmptySet()
-        expr_set_pairs = f.as_expr_set_pairs()
+        # the conditions can only be solved on the real domain
+        expr_set_pairs = f.as_expr_set_pairs(domain & S.Reals)
         for (expr, in_set) in expr_set_pairs:
             if in_set.is_Relational:
                 in_set = in_set.as_set()
-            if in_set.is_Interval:
-                dom -= in_set
             solns = solver(expr, symbol, in_set)
             result += solns
+    elif isinstance(f, Eq):
+        from sympy.core import Add
+        result = solver(Add(f.lhs, - f.rhs, evaluate=False), symbol, domain)
+    elif f.is_Relational:
+        if not domain.is_subset(S.Reals):
+            raise NotImplementedError(filldedent('''
+                Inequalities in the complex domain are
+                not supported. Try the real domain by
+                setting domain=S.Reals'''))
+        try:
+            result = solve_univariate_inequality(
+            f, symbol, domain=domain, relational=False)
+        except NotImplementedError:
+            result = ConditionSet(symbol, f, domain)
+        return result
```

Comment:
Still can't get why these have to get here. Can you give an example which would have failed otherwise?

Reply:
Any Piecewise that has a Relational expression, e.g. `solveset(Piecewise((x<1,x<5),(x<2,True)),x,S.Reals)` or `solveset(Piecewise((Eq(x - 1, 0), x<0), (Eq(x - 4, 0), True)), x, S.Reals)`.